File: week_3/3-advertisement_revenue/dot_product.hs
-- by Kirill Elagin

quicksort :: (Ord a) => [a] -> [a]
quicksort [] = []
quicksort (x:xs) = quicksort [y | y <- xs, y <= x] ++ [x] ++ quicksort [y | y <- xs, y > x]

acc :: [Int] -> Integer
acc [] = 0
acc (x:xs) = toInteger x + acc xs


max_dot_product :: [Int] -> [Int] -> Integer
max_dot_product as bs =
	let sort_as = quicksort as
	    sort_bs = quicksort bs 
	    mult_lst = zipWith (*) sort_as sort_bs 
	in acc mult_lst
-- result -- write your code here

main :: IO ()
main = do
  _  <- getLine
  as <- fmap (map read . words) getLine
  bs <- fmap (map read . words) getLine
  print $ max_dot_product as bs

File: week_3/3-advertisement_revenue/dot_product.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using std::vector;

long long max_dot_product(vector<long long> a, vector<long long> b)
{
  // write your code here
  std::sort(std::begin(a),std::end(a),std::greater<long long>());
  std::sort(std::begin(b),std::end(b),std::greater<long long>());

  std::transform(std::begin(a), std::end(a), std::begin(b), std::begin(a), std::multiplies<long long>());
  return std::accumulate(std::begin(a),std::end(a),0l);
}

int main() {
  size_t n;
  std::cin >> n;
  vector<long long> a(n), b(n);
  for (size_t i = 0; i < n; i++) {
    std::cin >> a[i];
  }
  for (size_t i = 0; i < n; i++) {
    std::cin >> b[i];
  }
  std::cout << max_dot_product(a, b) << std::endl;
}
